spec.md

```diff
@@ -538,6 +540,129 @@ It indicates the type of the data being stored inside of the memory.
+
+A circuit component is a **subcomponent** of another if there is way to get from the first component to the second through the direct subcomponent relation.
+A circuit component is trivially considered to be a subcomponent of itself.
+If we need to speak of the subcomponents excluding the component itself, we call these the **proper subcomponents**.
+
+A **root component** is a circuit component that is not the direct subcomponent of any other component.
+Equivalently, these are the circuit components which are declared inside a module.
+
+A **leaf component** is a circuit component has no direct subcomponents.
+Leaf components are useful for checking initialization.
+
+Two components are **disjoint** if they have no common subcomponent.
+For example, given `wire w : { x : UInt<1>, y : UInt<8>[2] }`{.firrtl}, `w.x`{.firrtl} and `w.y`{.firrtl} are disjoint, but `w.y[0]`{.firrtl} and `w.y`{.firrtl} are not.
+
+### References
+
+A **reference** is a name that refers to a previously declared circuit component.
+These are constructed using the names of root components, and optionally, through the dot and bracket operators.
+
+For instance, suppose we declare `wire w : { x : UInt<1>, v : UInt<8>[2] }`{.firrtl}.
+Then `w`{.firrtl}, `w.x`{.firrtl}, `w.y`{.firrtl}, `w.y[0]`{.firrtl}, and `w.y[1]`{.firrtl}, and `w.y[i]`{.firrtl} are all examples of references.
+
+When a circuit component has a vector type, you can perform both static and dynamic indexing.
+For example, given `wire v : UInt<8>[2]`{.firrtl}, both `v[0]`{.firrtl} and `v[1]`{.firrtl} are references using static indexing.
+On the other hand, `v[i]`{.firrtl} is a reference using dynamic indexing, given an expression `i`{.firrtl} with an unsigned integer type.
+
+The result of an out-of-bounds dynamic index is an indeterminate value (see [@sec:indeterminate-values]).
+
+A **static reference** is a reference where all indexing is static.
+For instances, `v[0]`{.firrtl}, `v[1]`{.firrtl}, and `v[2]`{.firrtl} are static references.
+A static reference identify a concrete subcomponent within the module definition.
+Static references may be used in `probe`{.firrtl} or `rwprobe`{.firrtl} expressions (see [@sec:probes]).
+
+A **dynamic reference** is one which is not a static reference.
+
+References may appear as the target of `connect`{.firrtl} statements (see [@sec:connections]).
+A connection to a dynamic reference is equivalent to a (potentially large) conditional statement where each branch connects to a static reference.
+For example this connection to a dynamic reference:
+
+``` firrtl
+  ; given  wire v : UInt<8>[4]
+  ; and    wire i : UInt<2>
+  connect v[i], UInt(0)
+```
+
+is equivalent to:
+
+``` firrtl
+  when eq(i, UInt(0)) :
+    connect v[0], UInt(0)
+  when eq(i, UInt(1)) :
+    connect v[1], UInt(0)
+  when eq(i, UInt(2)) :
+    connect v[2], UInt(0)
+  when eq(i, UInt(3)) :
+    connect v[3], UInt(0)
+  else :
+    invalidate v
+```
```

Comment:
This is not quite right. This either needs to have an `else : invalidate v` after every when or a single `invalidate v` needs to be moved before the first `when`. `v[0]`, `v[1]`, and `v[2]` are not properly initialized. This isn't normally important in examples like this. However, the intent is to show equivalence with the other code.